causalml/inference/tree/bart.py

```diff
@@ -0,0 +1,547 @@
+        return np.array(list(map(lambda z: max(z, 0), x.flatten()))).reshape(-1, 1)
+    
+    def fit(self, X:np.ndarray, y:np.ndarray, treatment:np.ndarray = None, max_iter:int = 100) -> None:
+        """Fit the BART for either uplift modeling or classic ML setting.
+        Args:
+            X (np.array): a feature matrix
+            y (np.array): an outcome vector
+            treatment (np.array): a treatment vector
+            max_ter (int): maximum number of iterations. Stopping criteria 
+        """
+
+        if treatment is not None:
+            self.problem_type = 'uplift_modeling'
+            if np.unique(treatment).size > 2:
+                raise Exception('Number of treatments greater than 2. Currently this method only supports binary treatment.')
+            self.treatments = np.unique(treatment).astype(int)
+            X = np.concatenate([X, treatment], axis=1)
+        else:
+            self.problem_type = 'classic_ml'
+            
+        self.task = 'regression' if len(np.unique(y)) > 2 else 'classification'
+        self.y_min = y.min()
+        self.y_max = y.max()
+        
+        if self.task == 'regression':
+            sigma_mu = 0.5 / (self.k * np.sqrt(self.m))
+            transformed_y = self.transform_y(y=y)
+            self.lambda_ = self.get_lambda(X, transformed_y)
+            alpha_ig = self.v / 2 
+            beta_ig = self.v * self.lambda_ / 2
+            sigma = self.sample_ig(alpha=alpha_ig, beta=beta_ig)
+        else:
+            transformed_y = y
+            sigma_mu = 3 / (self.k * np.sqrt(self.m))
+            sigma = 1
+            
+        mean_value_per_tree = transformed_y.mean() / self.m
+        self.sum_of_trees = transformed_y.mean()
+        self.trees = [Tree(root=Node(value=mean_value_per_tree, 
+                                     type_='terminal', 
+                                     data=Data(X=X, 
+                                               treatment=treatment,
+                                               y=transformed_y, 
+                                               R=transformed_y-mean_value_per_tree*(self.m-1),
+                                               index=np.array(list(range(len(X)))))), 
+                          ) for i in range(self.m)]
+    
+        for i in tqdm(range(max_iter), position=0, leave=True):
+            if self.task == 'regression':
+                response = transformed_y
+            else:
+                sample = np.random.normal(loc=self.sum_of_trees, scale=1, size=(len(transformed_y), 1))
+                max_els = self.max_el(sample)
+                min_els = self.min_el(sample)
+                response = np.where(y.reshape(-1, 1) == 1, max_els, min_els) # Z in the paper
+
+            for j in range(len(self.trees)):
+                residuals = self.compute_residual(X=X, y=response, trees=self.trees, j=j) # opportunity to parallelize according to Pratola et al. (2013) (https://arxiv.org/abs/1309.1906)
```

Comment:
Not sure what it says in the reference, but if I understand things correctly, for each tree we compute the residuals by looping over every other tree, which indeed seems wasteful. Is it not possible to store the prediction of each tree to avoid computing the same prediction over and over again?